drivers/sensors/sht4x.c

```diff
@@ -0,0 +1,827 @@
+ * Name: sht4x_crc_lookup
+ *
+ * Description:
+ *   Perform a CRC calculation using the lookup table.
+ *
+ ****************************************************************************/
+
+uint8_t sht4x_crc_lookup(const uint8_t *buf, uint8_t nbytes)
+{
+  uint8_t crc = SHT4X_CRC_INIT;
+  for (uint8_t byte = 0; byte < nbytes; byte++)
+    {
+      crc = CRC_LOOKUP[crc ^ buf[byte]];
+    }
+
+  return crc;
+}
+
+#endif
+
+/****************************************************************************
+ * Name: sht4x_crc_bitwise
+ *
+ * Description:
+ *   Perform a CRC calculation using the bitwise calculation.
+ *
+ ****************************************************************************/
+
+uint8_t sht4x_crc_bitwise(const uint8_t *buf, uint8_t nbytes)
+{
+  uint8_t crc = SHT4X_CRC_INIT;
+  for (uint8_t byte = 0; byte < nbytes; byte++)
+    {
+      crc ^= buf[byte];
+      for (uint8_t bit = 0; bit < 8; bit++)
+        {
+          if (crc & 0x80)
+            {
+              /* Discard the highest bit (implicit XOR), then divide by the
+               * polynomial.
+               */
+
+              crc <<= 1;
+              crc ^= SHT4X_CRC_POLY;
+            }
+          else
+            {
+              /* Continue until the highest bit is set. */
+
+              crc <<= 1;
+            }
+        }
+    }
+
+  return crc;
+}
+
+/****************************************************************************
+ * Name: sht4x_crc
+ *
+ * Description:
+ *   Perform a CRC calculation on the data to check for validity.
+ *   The `buf` param must be a data buffer that ends with its own 8 bit CRC.
+ *   The `nbytes` param is the length of the data including the 8 bit CRC.
+ *
+ ****************************************************************************/
+
+static int sht4x_crc(const uint8_t *buf, uint8_t nbytes)
+{
+#ifdef CONFIG_SHT4X_CRC_LOOKUP
+  if (sht4x_crc_lookup(buf, nbytes) != 0)
+    {
+      return -EBADMSG;
+    }
+#else
+  if (sht4x_crc_bitwise(buf, nbytes) != 0)
+    {
+      return -EBADMSG;
+    }
+
+#endif
+  return 0;
+}
+
+/****************************************************************************
+ * Name: sht4x_cmd
+ *
+ * Description:
+ *   I2C access helper for sending commands to the SHT4x which receive data
+ *   in response.
+ *
+ ****************************************************************************/
+
+static int sht4x_cmd(FAR struct sht4x_dev_s *priv, uint8_t command,
+                     uint32_t timeout, uint16_t *data1, uint16_t *data2)
+{
+  struct i2c_msg_s cmd;
+  struct i2c_msg_s read_data;
+  int err;
+
+  /* 2 byte data, 1 byte CRC, 2 byte data, 1 byte CRC. */
+
+  uint8_t combined_data[6];
+
+  cmd.frequency = CONFIG_SHT4X_I2C_FREQUENCY;
+  cmd.addr = priv->addr;
+  cmd.flags = 0;
+  cmd.buffer = &command;
+  cmd.length = sizeof(command);
+
+  /* Trigger measurement. */
+
+  err = I2C_TRANSFER(priv->i2c, &cmd, 1);
+  if (err) return err;
```

Comment:
Hi, I am definitely not the right person to do an review, I'm as much of a new contributor as you are.
But it happens to also develop an driver for an sensor, so I use your PR, and other sensors as guidelines.
Here if will return true on all non zero err values, even negative ones, but throughout other sensor drivers I see explicit check of non zero negative values `if (err < 0)`. Also, I see errors returned as strictly negative values. I know your driver behave as expected because the success/OK value is defined as 0.
Is this by design?

Reply:
Hello! Thanks for your review! That's a good point, I'm not sure of the reason why other drivers explicitly check for negative values. My reasoning was that only a '0' is success, so I want to return the error for all non-zero values. Maybe there's something I'm missing though, seeing as the other examples seem to be explicitly checking for negative values.